server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java

```diff
@@ -308,7 +315,17 @@ public final void addDynamicMapper(Mapper mapper) {
         // dynamically mapped objects when the incoming document defines no sub-fields in them:
         // 1) by default, they would be empty containers in the mappings, is it then important to map them?
         // 2) they can be the result of applying a dynamic template which may define sub-fields or set dynamic, enabled or subobjects.
-        dynamicMappers.add(mapper);
+        dynamicMappers.put(mapper.name(), mapper);
```

Comment:
What if the `mapper` is a different type than a previously stored one?

Reply:
Excellent catch! You're absolutely right about that. I think I've fixed it with 0f8029dc8b9d8ab413f7c91761e1a246c37343aa which also adds a test.

The reason I introduced the map and added `Mapper::mapperSize` is that the current implementation to count the number of additional fields via `newFieldsSeen.size()` was under-counting in case a new field has a multi-field. I added more details in a comment on the `getNewDynamicMappersSize` method.